File: packages/redsys/ios/Sources/RedsysPlugin/RedsysPlugin.swift
import Foundation
import Capacitor
import TPVVInLibrary

/**
 * Redsys Capacitor Plugin (iOS Bridge Layer)
 *
 * This class acts as the Bridge Layer between the JavaScript API
 * and the native Redsys iOS SDK.
 *
 * Responsibilities:
 * - Validate incoming JS parameters
 * - Forward calls to the Implementation layer
 * - Correlate SDK delegate callbacks with saved CAPPluginCall instances
 * - Map native errors to standardized JS-facing error codes
 *
 * Business logic MUST NOT live in this layer.
 */
@objc(RedsysPlugin)
public final class RedsysPlugin: CAPPlugin, CAPBridgedPlugin, DirectPaymentResponseDelegate, WebViewPaymentResponseDelegate {

    // MARK: - Properties

    /// Native implementation layer (contains SDK interaction logic).
    private let implementation = RedsysImpl()

    /// Read-only configuration loaded from capacitor.config.ts.
    private var config: RedsysConfig?

    /// Tracks the active direct payment call identifier
    private var activeDirectPaymentCallId: String?

    /// Internal Capacitor identifier.
    public let identifier = "RedsysPlugin"

    /// JavaScript-facing plugin name.
    public let jsName = "Redsys"

    /// Methods exposed to the JavaScript layer.
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "doDirectPayment", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "initializeWebPayment", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "processWebPayment", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "computeHash", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise)
    ]

    // MARK: - Lifecycle

    /**
     Called once when the plugin is loaded by Capacitor.

     Initializes configuration and injects it into the implementation layer.
     */
    override public func load() {
        // Initialize RedsysConfig with the correct type
        let cfg = RedsysConfig(plugin: self)
        self.config = cfg
        implementation.applyConfig(cfg)

        // Log if verbose logging is enabled
        RedsysLogger.debug("Plugin loaded.")
    }

    // MARK: - Error Handling

    /**
     Maps internal RedsysError cases to standardized JS error codes.
     */
    private func reject(
        _ call: CAPPluginCall,
        error: RedsysError
    ) {
        let code: String

        switch error {
        case .unavailable:
            code = "UNAVAILABLE"
        case .permissionDenied:
            code = "PERMISSION_DENIED"
        case .initFailed:
            code = "INIT_FAILED"
        case .unknownType:
            code = "UNKNOWN_TYPE"
        case .cryptoError:
            code = "CRYPTO_ERROR"
        }

        call.reject(error.message, code)
    }

    // MARK: - Direct Payment

    /**
     Starts the direct payment flow.

     Saves the call for asynchronous SDK callback correlation.
     */
    @objc func doDirectPayment(_ call: CAPPluginCall) {
        guard let order = call.getString("order") else {
            return call.reject("order required")
        }
        guard let amount = call.getDouble("amount") else {
            return call.reject("amount required")
        }

        // Save the call to correlate with delegate response via Ds_Order
        bridge?.saveCall(call)

        // Store callback identifier for KO correlation
        activeDirectPaymentCallId = call.callbackId

        implementation.executeDirectPayment(
            order: order,
            amount: amount,
            type: RedsysUtils.mapTransactionType(call.getString("transactionType")),
            description: call.getString("description"),
            identifier: call.getString("identifier"),
            extraParams: call.getObject("extraParams") as? [String: String],
            uiOptions: call.getObject("uiOptions"),
            delegate: self
        )
    }

    // MARK: - Web Payment Initialization

    /**
     Initializes WebView payment flow (Step 1 of 2-step 3DS process).
     */
    @objc func initializeWebPayment(_ call: CAPPluginCall) {
        guard let order = call.getString("order") else { return call.reject("order required") }
        // Changed getFloat to getDouble for parity with SDK and Android
        guard let amount = call.getDouble("amount") else { return call.reject("amount required") }

        implementation.initializeWebPayment(
            order: order,
            amount: amount,
            type: RedsysUtils.mapTransactionType(call.getString("transactionType")),
            description: call.getString("description"),
            identifier: call.getString("identifier"),
            extraParams: call.getObject("extraParams") as? [String: String]
        ) { base64, error in
            if let error = error as? RedsysError {
                self.reject(call, error: error)
            } else if let base64 = base64 {
                call.resolve(["base64Data": base64])
            }
        }
    }

    // MARK: - DirectPaymentResponseDelegate

    /**
     Called by the SDK when direct payment succeeds.
     */
    @objc public func responseDirectPaymentOK(response: DirectPaymentResponseOK) {
        if let call = bridge?.savedCall(withID: response.Ds_Order) {
            let result = RedsysUtils.mapDirectPaymentResponse(response)
            call.resolve(result)
            bridge?.releaseCall(call)
        }
    }

    /**
     * Called by the SDK when direct payment fails.
     * Fixed for Capacitor v8 iOS dictionary syntax.
     */
    @objc public func responseDirectPaymentKO(response: DirectPaymentResponseKO) {
        if let callbackId = activeDirectPaymentCallId,
           let call = bridge?.savedCall(withID: callbackId) {

            // In iOS, JSObject is a Dictionary, use standard subscripting
            var errorMetadata = [String: Any]()
            errorMetadata["sdkCode"] = response.code

            // Correct reject signature: message, code, error, data
            call.reject(
                "\(response.desc) (SDK Code: \(response.code))",
                "SDK_ERROR",
                nil, // No underlying Swift Error object
                errorMetadata
            )

            bridge?.releaseCall(call)
            activeDirectPaymentCallId = nil
        }
    }

    // MARK: - Web Payment Execution

    /**
     Executes WebView payment (Step 2 of 2-step flow).
     */
    @objc func processWebPayment(_ call: CAPPluginCall) {
        // Check call arguments first, then fallback to global config (RedsysConfig)
        let signature = call.getString("signature") ?? config?.signature

        guard let signature = signature, !signature.isEmpty else {
            return call.reject("Signature is required for web payment in call or configuration")
        }

        bridge?.saveCall(call)

        implementation.processWebPayment(
            signature: signature,
            signatureVersion: call.getString("signatureVersion"),
            delegate: self
        )
    }

    // MARK: - WebViewPaymentResponseDelegate

    /**
     Called by the SDK when WebView payment succeeds.
     */
    @objc public func responsePaymentOK(response: WebViewPaymentResponseOK) {
        if let call = bridge?.savedCall(withID: response.Ds_Order) {
            let result = RedsysUtils.mapWebViewResponse(response)
            call.resolve(result)
            bridge?.releaseCall(call)
        }
    }

    /**
     Called by the SDK when WebView payment fails.
     */
    @objc public func responsePaymentKO(response: WebViewPaymentResponseKO) {
        // WebView KO often uses code as identifier or generic
        if let call = bridge?.savedCall(withID: "\(response.code)") {
            var errorMetadata = [String: Any]()
            errorMetadata["sdkCode"] = response.code

            call.reject(
                "\(response.desc) (SDK Code: \(response.code))",
                "SDK_ERROR",
                nil,
                errorMetadata
            )
            bridge?.releaseCall(call)
        }
    }

    // MARK: - Cryptographic Utilities

    /**
     Computes HMAC signature using the provided data and key.
     */
    @objc func computeHash(_ call: CAPPluginCall) {
        guard let data = call.getString("data"),
              let key = call.getString("keyBase64") else {
            return call.reject("Missing data or key")
        }
        let alg = call.getString("algorithm") ?? "HMAC_SHA256_V1"

        if let signature = RedsysUtils.calculateHMAC(data: data, keyBase64: key, algorithm: alg) {
            call.resolve(["signature": signature])
        } else {
            // Precise error for debugging Base64 issues common in Redsys integration
            self.reject(call, error: .cryptoError("Hash computation failed. Check Base64 key integrity."))
        }
    }

    // MARK: - Version Information

    /**
     Returns the native plugin version.
     */
    @objc func getPluginVersion(_ call: CAPPluginCall) {
        call.resolve([
            "version": PluginVersion.number
        ])
    }
}

File: packages/redsys/android/src/main/java/io/capkit/redsys/RedsysPlugin.kt
package io.capkit.redsys

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.annotation.Permission
import com.redsys.tpvvinapplibrary.ErrorResponse
import com.redsys.tpvvinapplibrary.IPaymentResult
import com.redsys.tpvvinapplibrary.ResultResponse
import io.capkit.redsys.model.HashResult
import io.capkit.redsys.model.PluginVersionResult
import io.capkit.redsys.model.RedsysPaymentResponseOK
import io.capkit.redsys.model.RedsysWebPaymentInitResult
import io.capkit.redsys.utils.RedsysLogger
import io.capkit.redsys.utils.RedsysUtils
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import org.json.JSONObject

/**
 * Redsys Capacitor Plugin (Android Bridge Layer)
 *
 * Responsibilities:
 * - Validate incoming JavaScript parameters
 * - Delegate business logic to RedsysImpl
 * - Convert native results into JS-compatible objects
 * - Map native errors into standardized JS error codes
 *
 * Business logic MUST NOT live in this class.
 */
@CapacitorPlugin(
  name = "Redsys",
  permissions = [
    Permission(
      alias = "network",
      strings = [android.Manifest.permission.INTERNET],
    ),
  ],
)
class RedsysPlugin : Plugin() {
  // ---------------------------------------------------------------------------
  // Properties
  // ---------------------------------------------------------------------------

  /**
   * Immutable configuration container loaded from capacitor.config.ts.
   * Initialized once during plugin load.
   */
  private lateinit var config: RedsysConfig

  /**
   * Native implementation layer.
   * Contains SDK interaction and platform-specific logic.
   */
  private lateinit var implementation: RedsysImpl

  /**
   * Serializer instance configured to encode result models into JSObject string payloads.
   *
   * NOTE: encodeDefaults is intentionally NOT set so nullable/defaulted fields
   * are omitted from the emitted JSON, matching the optional TypeScript types.
   */
  private val json = Json

  // ---------------------------------------------------------------------------
  // Companion Object
  // ---------------------------------------------------------------------------

  private companion object {
    const val ACCOUNT_TYPE = "io.capkit.redsys"
    const val ACCOUNT_NAME = "Redsys"
  }

  // ---------------------------------------------------------------------------
  // Lifecycle
  // ---------------------------------------------------------------------------

  /**
   * Called once when the plugin is loaded by the Capacitor bridge.
   *
   * This is the correct place to:
   * - read static configuration
   * - initialize native resources
   * - inject configuration into the implementation
   */
  override fun load() {
    super.load()

    config = RedsysConfig(this)
    implementation = RedsysImpl(activity)
    implementation.updateConfig(config)

    RedsysLogger.verbose = config.verboseLogging
    RedsysLogger.debug("Plugin loaded. Version: ", BuildConfig.PLUGIN_VERSION)
  }

  // ---------------------------------------------------------------------------
  // Error Handling
  // ---------------------------------------------------------------------------

  /**
   * Maps internal RedsysError types to standardized JS error codes.
   */
  private fun reject(
    call: PluginCall,
    error: RedsysError,
  ) {
    val code =
      when (error) {
        is RedsysError.Unavailable -> "UNAVAILABLE"
        is RedsysError.PermissionDenied -> "PERMISSION_DENIED"
        is RedsysError.InitFailed -> "INIT_FAILED"
        is RedsysError.UnknownType -> "UNKNOWN_TYPE"
        is RedsysError.CryptoError -> "CRYPTO_ERROR"
      }

    call.reject(error.message, code)
  }

  /**
   * Converts SDK-level errors into a standardized JS error.
   * Preserves the original Redsys error code.
   */
  private fun rejectWithSdkError(
    call: PluginCall,
    error: ErrorResponse,
  ) {
    val data = JSObject()
    data.put("sdkCode", error.code)

    call.reject(
      error.desc,
      "SDK_ERROR",
      data,
    )
  }

  // ---------------------------------------------------------------------------
  // Serialization helpers
  // ---------------------------------------------------------------------------

  /**
   * Converts a serializable result model directly into a Capacitor JSObject.
   */
  private inline fun <reified T> toJSObject(value: T): JSObject {
    val jsonString = json.encodeToString(value)
    return JSObject(jsonString)
  }

  // ---------------------------------------------------------------------------
  // Direct Payment
  // ---------------------------------------------------------------------------

  /**
   * Starts the direct payment flow.
   * The result is delivered asynchronously via IPaymentResult.
   */
  @PluginMethod
  fun doDirectPayment(call: PluginCall) {
    val order = call.getString("order") ?: return call.reject("order required")
    val amount = call.getDouble("amount") ?: return call.reject("amount required")

    implementation.doDirectPayment(
      order = order,
      amount = amount,
      type = RedsysUtils.mapTransactionType(call.getString("transactionType")),
      id = call.getString("identifier"),
      desc = call.getString("description"),
      extra = RedsysUtils.toHashMap(call.getObject("extraParams")),
      uiOptions = call.getObject("uiOptions"),
      callback =
        object : IPaymentResult {
          override fun paymentResultOK(res: ResultResponse) {
            call.resolve(toJSObject(mapPaymentResult(res)))
          }

          override fun paymentResultKO(err: ErrorResponse) {
            rejectWithSdkError(call, err)
          }
        },
    )
  }

  // ---------------------------------------------------------------------------
  // Web Payment Initialization (Phase 1)
  // ---------------------------------------------------------------------------

  /**
   * Initializes WebView payment flow.
   * Returns Base64 merchant data for signature generation.
   */
  @PluginMethod
  fun initializeWebPayment(call: PluginCall) {
    val order =
      call.getString("order")
        ?: return call.reject("order required")

    val amount =
      call.getDouble("amount")
        ?: return call.reject("amount required")

    implementation.initializeWebPayment(
      order = order,
      amount = amount,
      type = RedsysUtils.mapTransactionType(call.getString("transactionType")),
      id = call.getString("identifier"),
      desc = call.getString("description"),
      extra = RedsysUtils.toHashMap(call.getObject("extraParams")),
      onResult = { base64 ->
        call.resolve(toJSObject(RedsysWebPaymentInitResult(base64Data = base64)))
      },
      onError = { error ->
        reject(call, error)
      },
    )
  }

  // ---------------------------------------------------------------------------
  // Web Payment Execution (Phase 2)
  // ---------------------------------------------------------------------------

  /**
   * Executes the WebView payment flow.
   */
  @PluginMethod
  fun processWebPayment(call: PluginCall) {
    // Using the 'signature' property from our RedsysConfig instance
    val signature = call.getString("signature") ?: if (config.signature.isNullOrEmpty()) null else config.signature

    if (signature == null) {
      return call.reject("signature required in call or configuration")
    }

    implementation.processWebPayment(
      signature,
      object : IPaymentResult {
        override fun paymentResultOK(res: ResultResponse) {
          call.resolve(toJSObject(mapPaymentResult(res)))
        }

        override fun paymentResultKO(err: ErrorResponse) {
          rejectWithSdkError(call, err)
        }
      },
    )
  }

  // ---------------------------------------------------------------------------
  // Cryptographic Utilities
  // ---------------------------------------------------------------------------

  /**
   * Computes HMAC signature for WebView payments.
   * WARNING: For production, signatures should be generated on the backend.
   */
  @PluginMethod
  fun computeHash(call: PluginCall) {
    val data = call.getString("data") ?: return call.reject("data required")
    val key = call.getString("keyBase64") ?: return call.reject("keyBase64 required")
    val alg = call.getString("algorithm") ?: "HMAC_SHA256_V1"

    // Redsys requires SHA256 or SHA512 depending on terminal config
    val signature = RedsysUtils.calculateHMAC(data, key, alg)

    if (signature != null) {
      call.resolve(toJSObject(HashResult(signature = signature)))
    } else {
      val error = RedsysError.CryptoError("Hash computation failed. Verify Base64 key format.")
      reject(call, error)
    }
  }

  // ---------------------------------------------------------------------------
  // Version Information
  // ---------------------------------------------------------------------------

  /**
   * Returns the native plugin version.
   */
  @PluginMethod
  fun getPluginVersion(call: PluginCall) {
    call.resolve(toJSObject(PluginVersionResult(version = BuildConfig.PLUGIN_VERSION)))
  }

  // ---------------------------------------------------------------------------
  // Result mapping
  // ---------------------------------------------------------------------------

  /**
   * Converts the SDK [ResultResponse] into the typed bridge result DTO
   * [RedsysPaymentResponseOK].
   *
   * This is the ONLY place that translates the SDK's generic response
   * into the bridge DTO, so the hand-rolled Map-to-JSObject conversion
   * in [RedsysUtils.resultToJSObject] is no longer needed.
   */
  private fun mapPaymentResult(res: ResultResponse): RedsysPaymentResponseOK {
    return RedsysPaymentResponseOK(
      code = res.responseCode?.toIntOrNull() ?: 0,
      desc = res.desc ?: "",
      amount = res.amount ?: "",
      currency = res.currency ?: "",
      order = res.order ?: "",
      merchantCode = res.merchantCode ?: "",
      terminal = res.terminal ?: "",
      responseCode = res.responseCode ?: "",
      authorisationCode = res.authorisationCode ?: "",
      transactionType = res.transactionType ?: "",
      securePayment = res.securePayment ?: "",
      signature = res.signature ?: "",
      cardNumber = if (res.cardNumber != null) RedsysUtils.maskCardNumber(res.cardNumber) else "",
      cardBrand = res.cardBrand ?: "",
      cardCountry = res.cardCountry ?: "",
      cardType = res.cardType ?: "",
      expiryDate = res.expiryDate ?: "",
      merchantIdentifier = res.identifier,
      consumerLanguage = res.language,
      date = res.date,
      hour = res.hour,
      merchantData = res.merchantData,
      extraParams = parseExtraParams(res.extraParams),
    )
  }

  /**
   * Parses the SDK's extraParams JSON string into a typed map.
   * Returns null when the input is null or empty.
   */
  private fun parseExtraParams(extraParams: String?): Map<String, String>? {
    if (extraParams.isNullOrEmpty()) return null
    return try {
      val json = JSONObject(extraParams)
      val map = mutableMapOf<String, String>()
      val keys = json.keys()
      while (keys.hasNext()) {
        val key = keys.next()
        map[key] = json.get(key).toString()
      }
      map
    } catch (_: Exception) {
      null
    }
  }
}